presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java

```diff
@@ -479,13 +480,15 @@ public void partitionPage(Page page)
             }
 
             // Decode the page just once. The decoded blocks will be fed to each PartitionBuffer object to set up AbstractBlockEncodingBuffer.
+            decodedBlocksRetainedSize = 0;
             for (int i = 0; i < decodedBlocks.length; i++) {
                 decodedBlocks[i] = decodeBlock(flattener, blockLeaseCloser, page.getBlock(i));
+                decodedBlocksRetainedSize += decodedBlocks[i].getRetainedSizeInBytes();
             }
 
             // Copy the data to their destination partitions and flush when the buffer is full.
             for (int i = 0; i < partitionBuffers.length; i++) {
-                partitionBuffers[i].appendData(decodedBlocks, page.getRetainedSizeInBytes(), fixedWidthRowSize, variableWidthChannels, outputBuffer);
+                partitionBuffers[i].appendData(decodedBlocks, decodedBlocksRetainedSize, fixedWidthRowSize, variableWidthChannels, outputBuffer);
```

Comment:
@yingsu00 I think moving away from page.getRetainedSizeInBytes() and using estimated flattened size is the right direction. It allows for much cleaner design. The DecodedBlockNode still needs the original value for the retainedSizeInBytes as it is used in memory tracking. Hence, I suggest to introduce a new variable to store the estimate of the flattened size of the block.